### DISCUSSION TITLE
Indexed views

### DISCUSSION THREAD

**Discussion by srid** [srid/emanote#50] [Ideas]
Neuron's [index](https://neuron.zettel.page/impulse) is just a tree of links to all notes. This works on small sized notebooks, but can become intractable on really large ones (and can affect performance adversely; see https://github.com/srid/emanote/issues/66). 

One solution to this problem is to *auto-collapse* sub-trees known to have relatively large number of descendants, and delegate their listing to a dedicated child (or sub) index route. For instance, daily notes stored in "Calendar/" folder could have as much as 300 notes in a year, and if you are using the Zettelkasten for 5 years, that number reaches 1500. It makes no sense to list all daily notes in the index view. We could just *automatically* collapse the Calendar/ sub-tree in the main index, and then introduce the notion of "sub index" views.

If the index view has the route `/-/index` (which is what #49 implements), then any sub-index view for path P will have the route `/-/index/$P`. For the above Calendar/ folder, the sub-index view will be at `/-/index/Calendar`. And for nested sub-index views, such as say a specific month of a year, it will be at `/-/index/Calendar/2020/07`.

---

We can in fact generalize this "indexed" route-tree concept and apply it to hierarchical tags too. So a note tagged with `foo/bar/qux` will be listed in the `/-/tags/foo/bar/qux` tag index route, with "qux" itself listed in `/-/tags/foo/bar` ([view demo](https://github.com/b0o/neuron-extras#demo)) and so on. 

---

- [x] Delineated Tag Index https://github.com/srid/emanote/pull/85
    - [ ] Add backlink context to tag index pages (display the paragraph in which the inline tag is used) #208 
- [ ] Delineated Folder Index (smart on-demand delineation?)
- [ ] Folgezettel Index (a la neuron's z-index)?
- [ ] #181

**Comment by srid** [comment]
Consolidate this with the sidebar tree configuration proposed in #64. See comment: https://github.com/srid/emanote/discussions/64#discussioncomment-921425

**Comment by wbadart** [comment]
I found this thread trying to track down what happened to the z-index in emanote. I definitely agree that neuron's z-index would not scale indefinitely, and that collapsible views could be part of the solution.

Thinking a little more aspirationally, maybe this is the opportunity to drop the tree view altogether and figure out some kind of graph visualization since back-links can form non-tree structures. We'd probably incur a frontend dependency to render the graph, but it gives way to a natural collapsing heuristic: hiding/ grouping nodes sharing parents of high out-degree.

**Reply by srid** [reply]
> Thinking a little more aspirationally, maybe this is the opportunity to drop the tree view altogether and figure out some kind of graph visualization

Yes, this is in my mind. Requires #84 first though ...

**Reply by wbadart** [reply]
Glad to see you're already tracking this. Thanks for the link :+1:

**Reply by srid** [reply]
I should clarify that the indexed view in its current form -- _a static structure of links: be it neuron's folgezettel heterarchy dump, or a collapsible version being proposed here, or the full directory index of emanote in its current form_ -- is still useful, and should be retained. Dynamic graph view can be provided *in addition to* it. Only when the graph views satisfies all the use cases of indexed view may the latter be deprecated (otherwise, both should be kept). 

I think any work on graph view would be in a very early research and exploration phase (I'm pretty new to visualization myself), so I don't know how it will turn out yet. The index view should be a good thing to have in the interim. I do know that I'm not a big fan of Obsidian's graph view, and as @j-steinbach notes below it is not a great index view (just that it looks pretty) ... but perhaps there is a way we can improve on that in an interactive fashion including #147 based interaction so as to provide a way to iteratively 'filter' the graph in real-time based on one's (changing) query view of what notes need to be displayed. All of this sounds like an independent project of its own, and maybe it should be explored in a separate repo first (will require #161).

**Comment by j-steinbach** [comment]
I believe the big question is: What exactly do we mean by index? Is it to display _all_ notes? To display the _important_ notes? To show an overview over collapsed categories? Also: What is its purpose? Look pretty? Provide a sense of pride and accomplishment by showing how big our notes folder is? Be useful? Let the user look up their notes-folder from the browser?

[Side-note: Our [index](https://neuron.zettel.page/impulse) is called "Impulse" -- "Full folgezettel heterarchy of the zettelkasten". What does that mean? Also we got a [index.html](https://neuron.zettel.page/) page, which is user curated.]

If we assume our index to be intended as a listing of all notes, then we have the following problem: Our notes folder has no structure. We can't assume structure, and we mustn't enforce structure. It is just a loose pile of notes, spread over a few (possibly none) folders. 

For example my Zettelkasten (ZK) consists of five folders:

- _root_ :: ~2000 loose notes
- _root/science_ :: ~1500 notes + bibliography
- _root/literate_ :: my [literate](https://github.com/zyedidia/Literate) system config. ~100 files
- _root/writings_ :: my novel + world-building ~100 files
- _root/diary_ :: two years worth of daily notes (~700)

So how do we structure this?

- List all files :: This is what we currently [do](https://neuron.zettel.page/impulse) -- called "Full folgezettel heterarchy of the zettelkasten." Problem: A list of 2000 loose files doesn't help anyone.
- List all files, only show the parent nodes. :: This is what the [visualization](https://assets.niklas-luhmann-archiv.de/branchview/#ZK_1_NB_7-2m_V) of the original ZK does. The index on the left only show the parent nodes, and we can explore the children. Problem: How do we deal with loose notes? Just list them under a category "Orphans"?
- List files based on their ids. Problem: This assumes https://github.com/srid/emanote/discussions/142
- List files alphabetically (Collapse them similar to Wikimedia [subcategories](https://commons.wikimedia.org/wiki/Category:Earth)) Problem: Internationalization. Example: What is the alphabetical order of [CJK](https://en.wikipedia.org/wiki/CJK_characters) notes?
- Categorize based on tags. Similar to Wikimedia sidebar, under ["Content"](https://commons.wikimedia.org/wiki/Main_Page). Problem: Is this still an index? Also we might have thousands of tags again (multi-lingual).

This means that a graph probably is the safest way to visualize the notes. But is a graph an index? In my opinion a graph is a visualization, which --whilst pretty and awesome-- doesn't have much "index" value:

![zettelkasten](https://user-images.githubusercontent.com/69524139/135753620-b3ebec4e-14fd-4682-84ae-b776176f027b.png)

The picture shows a cut-out of my Zettelkasten graph. The real one is 4x bigger and also contains notes without parents and links. I find it hard to identify small notes. **The problem with graphs is that they change**. In my opionion, an index should show me the most important parents/categories on a single glance, without changing every second day or requiring me to zoom around.

This does not mean that I am against graphs. _Au contraire!_, we definitely need a graph-based visualization. I just think that graphs and indices are different things.

**Reply by Terkwood** [reply]
Personally I don't love the idea of the folgezettel hierarchy display going away. It is easier for me to read than a simple graph view as described by @j-steinbach. I also keep notes in the low thousands and the predictable visualization of neuron style or current ema style is helpful to me.

**Reply by wbadart** [reply]
The folgezettel hierarchy is also my main entrypoint into my notes when I'm looking for something specific. The main problem I have with the current layout is that notes which fall under multiple "branches" of the heterarchy are only shown under one (with the link icon indicating that they're referenced in other branches/ categories).

Using an example from my own (very small) ZK:

![Screenshot 2021-10-03 at 10-02-40 Impulse](https://user-images.githubusercontent.com/8717474/135764095-6e221449-d3ae-4c7e-99e8-f3938fcffc50.png)

Here is the view of all notes which have the `Statistics` zettel as an ancestor. As you can see, several of them have the little link icon, because they also have `Computer Science` as an ancestor. If I were looking through my `Computer Science` notes for a probabilistic programming language (e.g. for the `Stan` zettel), I wouldn't see it, as it would be hidden under the collapsed `Statistics` section.

Personally I'd prefer to have it duplicated, and appear under both `Computer Science` and `Statistics`, though I recognize I may be in the minority here.

**Reply by j-steinbach** [reply]
From a logical perspective I don't see why we should show a note in the first category (statistics) and then not in the second one (CS). 

That means we have to list entries multiple times (assuming we want to recognize "top-level" notes as important and want to show them and their "descendant tree"). We* might circumvent the showing of multiple children in the tree structure by implementing something like [haxplore](https://haxplore.com/). Imagine the first column as our important top-level notes (notes that have no parents themselves). The second column shows their children, and so on. We can easily traverse the tree and also circumvent the issue of multiple nodes: They appear in separate columns, so they can be reused without cluttering our view. Nevertheless, the problem of "too many notes" still stands, mainly in the form of orphans (no parents, no children), but we might solve this by aggregating important notes (have many children) to the top. 

A quick mockups to demonstrate the core idea -- but I am not conviced that the note should show a preview. It probably should only show its number of descendants, and maybe creation date and/or tags.

[Bonus thought: Icons for notes?]

![emanote-index](https://user-images.githubusercontent.com/69524139/135768147-e901c8c7-8cd5-405a-b550-c19f1423e64a.png)
*Always nice to speak as a group if you are not the one doing the work :smile:

**Reply by almereyda** [reply]
Your mockup reminds me of two things in particular:

- Gingko looks very similar to what you explore, and is easy to browse, scroll through with the arrow keys.  
  ![gingko-big-picture-details](https://user-images.githubusercontent.com/1645308/135935440-39fb9c2d-20b2-47fe-96bb-b5a628a83685.png)
  * [Gingko App](https://gingkoapp.com/)
  * [Gingko Writer](https://gingkowriter.com/)
  * [Gingko Inc.  · GitHub](https://github.com/gingko)
  * [Releases · gingko/client · GitHub](https://github.com/gingko/client/releases)
- the [Federated Wiki Title Network Browser](http://ward.asia.wiki.org/title-network-browser.html) uses a similar column layout to display two levels of each incoming (to the left) and outgoing (to the right) links.  
  ![Screenshot 2021-10-05 at 00-57-04 Screenshot](https://user-images.githubusercontent.com/1645308/135935658-02e6d179-efe1-4536-8b6b-df2c8b379b08.png)  
  The numbers in this image visualise something very specific to the wiki federation:
  > We sum the forks for each title and then show this and the title count as a column heading. Dividing these two numbers gives you the average degree of duplication.

**Reply by j-steinbach** [reply]
Yes, this is exactly what I meant. Simple to explore, high information-density and an elegant way to handle duplicate entries.

(Well, the Federated Wiki Title Network Browser however is a bit too dense on information for my tastes :smile: )

**Reply by dpaetzel** [reply]
I just wanted to ask whether by now there are any plans to incorporate indexed views á la neuron (or similar, some of the above proposals seem very promising as well) into emanote? Right now, this is the main reason that's holding me back from switching to emanote.

Anyway, thank you for your work, @srid ! :slightly_smiling_face:

**Reply by srid** [reply]
@dpaetzel No plans so far, but I'm happy to guide others if they want to implement any of it. I can imagine two parts: adding extra route types (if necessary) and adding a 'render' function - which is a pure function taking 'Model' and returning HTML, effectively. Like, for tag index, we have:

- Route: https://github.com/srid/emanote/blob/6f83b857741fb6607716cf97cee3382c7f029180/src/Emanote/Route/SiteRoute/Type.hs#L38
- render: https://github.com/srid/emanote/blob/6f83b857741fb6607716cf97cee3382c7f029180/src/Emanote/View/TagIndex.hs#L79-L80

**Reply by dpaetzel** [reply]
OK, thanks for pointing us in the general direction. When I find the time (though I'm not sure when that may be), I might give it a go.

**Comment by srid** [comment]
**Note** that Emanote now has [JSON export](https://emanote.srid.ca/-/export.json); so these views can experimented with offline outside of Emanote before being upstreamed.

**Comment by srid** [comment]
I'm beginning to brainstorm over this. If anybody is interested, join the discussion [over at Zulip](https://nixos.zulipchat.com/#narrow/stream/413951-general/topic/obsidian.20graph.20in.20emanote).
